File: web/src/app/store/domain/log.ts
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as yaml from 'js-yaml';
import { LogType, Severity } from 'src/app/store/domain/style';
import { LogStore } from 'src/app/store/domain/log-store';
import { ReadonlyDomainElement } from 'src/app/store/domain/types';
import { BigIntTimeUtil } from 'src/app/utils/bigint-time-util';

/**
 * Lazy adapter for a Log entry.
 * Evaluates fields on demand and ensures deep immutability.
 */
export class Log {
  constructor(
    public readonly id: number,
    private readonly store: LogStore,
  ) {}

  /**
   * Gets the chronological index of this log in the store.
   */
  get logIndex(): number {
    return this.store.getIndex(this.id);
  }

  /**
   * Gets the timestamp of the log entry.
   */
  get timestamp(): bigint {
    return this.store._getTimestamp(this.id);
  }

  /**
   * Gets the timestamp (in milliseconds) of the log entry.
   * @deprecated Use {@link timestamp} instead, which returns the timestamp in nanoseconds.
   */
  get legacyTimestampMs(): number {
    return BigIntTimeUtil.NsToNumberMs(this.timestamp);
  }

  /**
   * Gets the human-readable summary of the log.
   */
  get summary(): string {
    return this.store._getSummary(this.id);
  }

  /**
   * Gets the associated log source category.
   */
  get logType(): ReadonlyDomainElement<LogType> {
    return this.store._getLogType(this.id);
  }

  /**
   * Gets the severity priority of the log.
   */
  get severity(): ReadonlyDomainElement<Severity> {
    return this.store._getSeverity(this.id);
  }

  /**
   * Gets the structured log attributes decoded from Intern pool data stores.
   */
  get body(): ReadonlyDomainElement<Record<string, unknown>> | null {
    return this.store._decodeBody(this.id);
  }

  /**
   * Gets the YAML string representation of the structured log attributes.
   */
  get bodyYAML(): string {
    return this.body ? yaml.dump(this.body, { lineWidth: -1 }) : '';
  }
}

File: pkg/model/log/log.go
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package log

import (
	"time"

	"github.com/GoogleCloudPlatform/khi/pkg/common/structured"
	"github.com/GoogleCloudPlatform/khi/pkg/model/id"
)

// Log represents a log handled in KHI.
// It provides direct access to its fields and basic metadata such as timestamp and instance ID.
type Log struct {
	*structured.NodeReader
	Timestamp time.Time
	ID        uint32
}

// NewLog returns a log instance from NodeReader instance using the given ID Generator.
func NewLog(gen *id.Generator, reader *structured.NodeReader) *Log {
	return &Log{
		ID:         gen.New(id.TemporaryLog),
		NodeReader: reader,
	}
}

// NewLogWithTimestamp returns a log instance with the given NodeReader and timestamp using the given ID Generator.
func NewLogWithTimestamp(gen *id.Generator, reader *structured.NodeReader, timestamp time.Time) *Log {
	return &Log{
		ID:         gen.New(id.TemporaryLog),
		NodeReader: reader,
		Timestamp:  timestamp,
	}
}

// NewLogFromYAMLString instantiates a new Log from the given YAML string using the given ID Generator.
func NewLogFromYAMLString(gen *id.Generator, yaml string) (*Log, error) {
	node, err := structured.FromYAML(yaml)
	if err != nil {
		return nil, err
	}
	return NewLog(gen, structured.NewNodeReader(node)), nil
}
